File: backend/src/Bojan.Application/Contracts/InvoiceSettings.cs
using System.Text.Json;

namespace Bojan.Application.Contracts;

/// <summary>Who the invoice says is selling — the block screen 34 draws opposite the buyer.</summary>
public sealed record InvoiceSellerDto(
    string Name,
    string Website,
    string Email,
    string Phone,
    string Address,
    /// <summary>شناسه ملی — printed only when the shop has one to print.</summary>
    string NationalId,
    /// <summary>کد اقتصادی — same.</summary>
    string EconomicCode);

/// <summary>
/// The parts of the invoice that are the shop's words rather than the order's
/// facts.
/// </summary>
/// <remarks>
/// <para>
/// Everything here was hard-coded in the document component, which made the
/// shop's own legal footer a code change. It is the owner's copy: a seller
/// address, a support address, the terms a buyer is told they accepted. None of
/// it is derivable from an order, and all of it changes without a deploy.
/// </para>
/// <para>
/// Served on the invoice itself rather than fetched separately by whoever draws
/// it, so the customer's copy and the panel's cannot be rendered from different
/// settings — the same reason both read one <see cref="InvoiceDto"/>.
/// </para>
/// </remarks>
public sealed record InvoiceSettingsDto(
    InvoiceSellerDto Seller,
    /// <summary>The line above the terms — "از اعتماد و خرید شما سپاسگزاریم."</summary>
    string ThanksNote,
    string Terms,
    /// <summary>The footer line under the rule.</summary>
    string FooterNote,
    /// <summary>
    /// The uploaded electronic stamp, or null.
    /// </summary>
    /// <remarks>
    /// Null is not a missing value to fill in with a default: there is no
    /// default stamp, and an invented one on a document that settles money
    /// would be a forgery rather than a placeholder. The document draws an
    /// empty box to stamp by hand until a real file is uploaded.
    /// </remarks>
    string? StampUrl)
{
    /// <summary>The settings section these live under.</summary>
    public const string Section = "invoice";

    /// <summary>
    /// What the invoice says when nothing has been configured.
    /// </summary>
    /// <remarks>
    /// A shop that has never opened the settings screen still prints a complete
    /// document — the same rule the cancellation percentage follows. These are
    /// the values the component carried before they were configurable.
    /// </remarks>
    public static InvoiceSettingsDto Defaults { get; } = new(
        new InvoiceSellerDto(
            Name: "فروشگاه بوژان",
            Website: "bojanstore.com",
            Email: "support@bojanstore.com",
            Phone: string.Empty,
            Address: string.Empty,
            NationalId: string.Empty,
            EconomicCode: string.Empty),
        ThanksNote: "از اعتماد و خرید شما سپاسگزاریم.",
        Terms: "خریدار با ثبت این سفارش، قوانین و مقررات فروشگاه بوژان را مطالعه کرده و پذیرفته است. "
             + "مهلت مرجوعی کالا و شرایط گارانتی، مطابق شرایط اعلام‌شده در زمان خرید است.",
        FooterNote: "این فاکتور به‌صورت الکترونیکی صادر شده و بدون مهر و امضای فیزیکی نیز معتبر است.",
        StampUrl: null);

    /// <summary>
    /// Builds the settings from the stored rows, falling back per field.
    /// </summary>
    /// <param name="stored">
    /// Section rows as <c>key -&gt; JSON</c>. Values are JSON-encoded strings,
    /// quotes included — see <c>SettingEntry.Value</c>.
    /// </param>
    /// <remarks>
    /// Per field rather than all-or-nothing: an owner who set only a support
    /// address must not lose the rest of the document to blanks, and a key this
    /// version does not know about is ignored rather than fatal.
    /// </remarks>
    public static InvoiceSettingsDto From(IReadOnlyDictionary<string, string> stored)
    {
        string Read(string key, string fallback)
        {
            if (!stored.TryGetValue(key, out var raw) || string.IsNullOrWhiteSpace(raw))
            {
                return fallback;
            }

            try
            {
                // Written by the panel as JSON. A row hand-edited to bare text
                // is read as that text rather than throwing — the settings
                // table is not a place to be brittle about quoting.
                var value = JsonSerializer.Deserialize<string>(raw);
                return string.IsNullOrWhiteSpace(value) ? fallback : value;
            }
            catch (JsonException)
            {
                return raw;
            }
        }

        var defaults = Defaults;

        return new InvoiceSettingsDto(
            new InvoiceSellerDto(
                Read("sellerName", defaults.Seller.Name),
                Read("sellerWebsite", defaults.Seller.Website),
                Read("sellerEmail", defaults.Seller.Email),
                Read("sellerPhone", defaults.Seller.Phone),
                Read("sellerAddress", defaults.Seller.Address),
                Read("sellerNationalId", defaults.Seller.NationalId),
                Read("sellerEconomicCode", defaults.Seller.EconomicCode)),
            Read("thanksNote", defaults.ThanksNote),
            Read("terms", defaults.Terms),
            Read("footerNote", defaults.FooterNote),
            // The one field with no fallback: absent means no stamp, and the
            // document draws the empty box instead.
            Read("stampUrl", string.Empty) is { Length: > 0 } stamp ? stamp : null);
    }
}

File: frontend/apps/admin/src/lib/api/invoice-settings.ts
import { api, useMockData } from './client';
import { INVOICE_KEYS, INVOICE_SECTION } from '@/lib/invoice-fields';
import type { InvoiceSettingsDto } from './types';

// Re-exported so server callers have one import for the whole thing. The
// definitions live in a module with no imports of its own, because the client
// form needs them too and must not pull the API client in with them.
export { INVOICE_KEYS, INVOICE_SECTION };

/**
 * What the invoice says when nothing has been configured.
 *
 * A copy of the API's `InvoiceSettingsDto.Defaults`. Duplicated rather than
 * fetched because this is what the *form* shows an owner who has never saved —
 * the document itself never reads this file, it reads the settings the API
 * resolved onto the invoice. The two agreeing matters only so the form does not
 * promise something different from what will print.
 */
export const INVOICE_DEFAULTS: InvoiceSettingsDto = {
  seller: {
    name: 'فروشگاه بوژان',
    website: 'bojanstore.com',
    email: 'support@bojanstore.com',
    phone: '',
    address: '',
    nationalId: '',
    economicCode: '',
  },
  thanksNote: 'از اعتماد و خرید شما سپاسگزاریم.',
  terms:
    'خریدار با ثبت این سفارش، قوانین و مقررات فروشگاه بوژان را مطالعه کرده و پذیرفته است. مهلت مرجوعی کالا و شرایط گارانتی، مطابق شرایط اعلام‌شده در زمان خرید است.',
  footerNote: 'این فاکتور به‌صورت الکترونیکی صادر شده و بدون مهر و امضای فیزیکی نیز معتبر است.',
  stampUrl: null,
};


/**
 * The saved invoice settings, with each unset field falling back to its
 * default.
 *
 * Per field rather than all-or-nothing, exactly as the API resolves them: an
 * owner who set only a support address must not find the rest of the form
 * blank.
 */
export async function getInvoiceSettings(): Promise<InvoiceSettingsDto> {
  if (useMockData) return INVOICE_DEFAULTS;

  const stored = await api
    .get<Record<string, string>>(`/settings/${INVOICE_SECTION}`, { auth: true })
    .catch(() => ({}) as Record<string, string>);

  const read = (key: string, fallback: string) => {
    const raw = stored[key];
    return raw === undefined || raw.trim() === '' ? fallback : unquote(raw);
  };

  const stamp = stored[INVOICE_KEYS.stampUrl];

  return {
    seller: {
      name: read(INVOICE_KEYS.sellerName, INVOICE_DEFAULTS.seller.name),
      website: read(INVOICE_KEYS.sellerWebsite, INVOICE_DEFAULTS.seller.website),
      email: read(INVOICE_KEYS.sellerEmail, INVOICE_DEFAULTS.seller.email),
      phone: read(INVOICE_KEYS.sellerPhone, ''),
      address: read(INVOICE_KEYS.sellerAddress, ''),
      nationalId: read(INVOICE_KEYS.sellerNationalId, ''),
      economicCode: read(INVOICE_KEYS.sellerEconomicCode, ''),
    },
    thanksNote: read(INVOICE_KEYS.thanksNote, INVOICE_DEFAULTS.thanksNote),
    terms: read(INVOICE_KEYS.terms, INVOICE_DEFAULTS.terms),
    footerNote: read(INVOICE_KEYS.footerNote, INVOICE_DEFAULTS.footerNote),
    // No default: absent means no stamp, and the document draws an empty box
    // to stamp by hand rather than inventing artwork.
    stampUrl: stamp && stamp.trim() !== '' ? unquote(stamp) : null,
  };
}

/**
 * Settings rows are stored as written, but a row saved by an older build — or
 * by hand — may carry JSON quotes. Stripping them here means the form shows the
 * value rather than the value wrapped in quotation marks.
 */
function unquote(raw: string): string {
  if (raw.length >= 2 && raw.startsWith('"') && raw.endsWith('"')) {
    try {
      return JSON.parse(raw) as string;
    } catch {
      return raw;
    }
  }
  return raw;
}
